system/src/system_task.cpp

```diff
@@ -347,79 +387,100 @@ int cloud_handshake()
+// TODO: Move this function to system_cloud_internal.cpp
 void handle_cloud_connection(bool force_events)
 {
-    if (SPARK_CLOUD_SOCKETED)
-    {
-        if (!SPARK_CLOUD_CONNECTED && !SPARK_CLOUD_HANDSHAKE_PENDING)
-        {
-            int err = 0;
-            if (SPARK_CLOUD_HANDSHAKE_NOTIFY_DONE) {
-                // TODO: There's no protocol API to get the current session state, so we're running
-                // one more iteration of the communication loop to make sure all handshake messages
-                // have been acknowledged successfully
-                if (!Spark_Communication_Loop()) {
-                    err = particle::protocol::MESSAGE_TIMEOUT;
+    if (!SPARK_CLOUD_SOCKETED) {
+        return;
+    }
+    if (!SPARK_CLOUD_CONNECTED) {
+        int err = 0;
+        if (g_cloudHandshakeState == CloudHandshakeState::PENDING) {
+            LED_SIGNAL_START(CLOUD_HANDSHAKE, NORMAL);
+            err = cloud_handshake();
+            if (err == 0) {
+                if (hasPendingClientMessages()) {
+                    g_cloudHandshakeState = CloudHandshakeState::WAIT_HANDSHAKE_ACKS;
+                    LOG(TRACE, "Waiting ACKs for protocol/application handshake messages");
+                } else {
+                    g_cloudHandshakeState = CloudHandshakeState::SEND_COMPLETE;
+                }
+            }
+        }
+        if (g_cloudHandshakeState == CloudHandshakeState::SEND_COMPLETE) {
+            const auto p = spark_protocol_instance();
+            err = spark_protocol_command(p, ProtocolCommands::HANDSHAKE_COMPLETE);
+            if (err == 0) {
+                if (hasPendingClientMessages()) {
+                    g_cloudHandshakeState = CloudHandshakeState::WAIT_COMPLETE_ACK;
+                    LOG(TRACE, "Waiting ACK for HandshakeComplete message");
                 } else {
-                    INFO("Cloud connected");
-                    SPARK_CLOUD_CONNECTED = 1;
-                    SPARK_CLOUD_HANDSHAKE_NOTIFY_DONE = 0;
-                    cloud_failed_connection_attempts = 0;
-                    CloudDiagnostics::instance()->status(CloudDiagnostics::CONNECTED);
-                    system_notify_event(cloud_status, cloud_status_connected);
-                    if (system_mode() == SAFE_MODE) {
-/* FIXME: there should be macro that checks for NetworkManager availability */
-                        // Connected to the cloud while in safe mode
+                    g_cloudHandshakeState = CloudHandshakeState::DONE;
+                }
+            } else if (err == protocol::ProtocolError::NOT_IMPLEMENTED) { // Core?
+                g_cloudHandshakeState = CloudHandshakeState::DONE;
+                err = 0;
+            }
+        }
+        if (g_cloudHandshakeState == CloudHandshakeState::DONE) {
+            // TODO: There's no protocol API to get the current session state, so we're running
+            // one more iteration of the communication loop to make sure all handshake messages
```

Comment:
I am kind of puzzled by this one too!

Reply:
This is not a new code, but, yeah, the comment could be more clear. The problem is that when a CoAP message expires, the event loop function doesn't return an error (instead, the protocol implementation [closes the channel](https://github.com/particle-iot/device-os/blob/2a8f892646022f50f9e5572a8132811961ed6111/communication/src/coap_channel.cpp#L57-L63) silently), and we have to invoke the loop function one more time to determine if the current protocol session is still valid.